lib/db_connection.ex

```diff
@@ -785,65 +894,382 @@ defmodule DBConnection do
+  implementation.
+
+  ### Options
+
+    * `:pool_timeout` - The maximum time to wait for a reply when making a
+    synchronous call to the pool (default: `5_000`)
+    * `:queue` - Whether to block waiting in an internal queue for the
+    connection's state (boolean, default: `true`)
+    * `:timeout` - The maximum time that the caller is allowed the
+    to hold the connection's state (ignored when using a locked connection,
+    default: `15_000`)
+    * `:log` - A function to log information about a call, either
+    a 1-arity fun, `{module, function, args}` with `DBConnection.LogEntry.t`
+    prepended to `args` or `nil`. See `DBConnection.LogEntry` (default: `nil`)
+
+  The pool and connection module may support other options. All options
+  are passed to `handle_commit/2`.
+
+  See `begin/2` and `rollback/2`.
+
+  ### Example
+
+      {:ok, result} = DBConnection.begin(conn)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.commit(conn)
+      end
+  """
+  @spec commit(conn, opts :: Keyword.t) :: {:ok, result} | {:error, Exception.t}
+  def commit(conn, opts \\ []) do
+    conn
+    |> run(&run_commit/4, meter(opts), opts)
+    |> log(:commit, :commit, nil)
+  end
+
+  @doc """
+  Commit a transaction.
+
+  Returns `result` on success, otherwise raises an exception on error.
+
+  See `commit/2`.
+  """
+  @spec commit!(conn, opts :: Keyword.t) :: result
+  def commit!(conn, opts \\ []) do
+    case commit(conn, opts) do
+      {:ok, result} ->
+        result
+      {:error, err} ->
+        raise err
+    end
+  end
+
+  @doc """
+  Execute a commit request, and if successful release the lock on connection.
+
+  Return `{:ok, result}` on success or `{:error, exception}` if there was an
+  error. On error the connection is not released back to the pool.
+
+  It is possible to issue multiple commit requests without a begin (`begin/2`
+  or `checkout_begin/2`). The semantics are left to the callback
+  implementation.
+
+  ### Options
+
+    * `:log` - A function to log information about a call, either
+    a 1-arity fun, `{module, function, args}` with `DBConnection.LogEntry.t`
+    prepended to `args` or `nil`. See `DBConnection.LogEntry` (default: `nil`)
+
+  The pool and connection module may support other options. All options
+  are passed to `handle_commit/2`.
+
+  See `checkout_begin/2` and `rollback_checkin/2`.
+
+  ### Example
+
+      {:ok, conn, result} = DBConnection.checkout_begin(pool)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.commit_checkin(conn)
+      end
+  """
+  @spec commit_checkin(t, opts :: Keyword.t) ::
+    {:ok, result} | {:error, Exceptiont.t}
+  def commit_checkin(%DBConnection{} = conn, opts) do
+    return =
+      with {:ok, result, meter}
+            <- run(conn, &run_commit/4, meter(opts), opts) do
+        checkin(conn, opts)
```

Comment:
Define fail? `checkin/2` will not raise, but it may return an `{:error, %DBConnection.Error{}}` if the socket is disconnected. For the socket to be disconnect at this moment the commit must have disconnected, and for that happen the `with` pattern doesn't match and `checkin/2` isn't called. We are removing this function though because it is not easy to understand.